### PR TITLE
Fixes #704: Added Config.py

### LINKED ISSUE #647: Seperating Config from Code base
**I'm submitting a ...**
- [x] bug report
- [ ] feature request

**Parent Issue**: https://github.com/fossasia/badgeyay/issues/627

**Current behavior:**
Currently all the configs are present in one single file

**Expected behavior:**
`config.py` should be present in the codebase where all the config should be managed.

**Yes, i want to work on this issue**

### LINKED ISSUE #704: Config.py is not present
**I'm submitting a ...**
- [x] bug report
- [ ] feature request

**Current behavior:**

Currently, we have to copy `dev-config.py` and make new file `config.py` and update configuration

**Expected behavior:**

`Config.py` should be present and only `.env` file should be changed

### PR DESCRIPTION
**Pull request by vaibhavsingh97**
Fixes #704 and Fixes #647

#### Checklist
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:
- Added .env file to store configurations
- Updated requirements

### CODE REVIEW

**Comment by vaibhavsingh97** [comment]
@ParthS007 @gabru-md @yashLadha @djmgit Please review. I have done all the requested changes

**Comment by vaibhavsingh97** [comment]
@gabru-md I had done the requested changes. Please review

**Comment by djmgit** [comment]
@yashLadha please have a look

**Comment by djmgit** [comment]
@ParthS007 @gabru-md @yashLadha please review this PR so that we can speed up the development process.

**Comment by gabru-md** [comment]
There are hound errors that need to be resolved.

**Comment by vaibhavsingh97** [comment]
@gabru-md according to hound, `create_app` is undefined but I already imported. Please have a look in the source code. Thanks

**Comment by harshit98** [comment]
Looks good to me. @vaibhavsingh97 is right. `create_app` is already there. @ParthS007 @yashLadha 
 please review this PR so that it can be merged and as @djmgit pointed out you guys have to speed the development process to meet the project outcomes.

It can be only possible:

1. Working together and being a good team player. ✨ 
2. Open PRs if and only if feature has been implemented completely. I mean avoid PRs with labels 'work in progress' or 'do not merge'. ⚠️ 
3. Ask mentors on the specific channel for help/guidance if stuck somewhere.
4. Avoid arguments. Arguments can be technical based with supporting resources. Respect each-other opinions.

Please let me know or @djmgit if you guys face any development related issues during GSoC.

File: backend/blueprint/api/run.py [hunk 1]
```diff
@@ -1,9 +1,7 @@
 import os
-from flask import Flask
-from flask_migrate import Migrate
-
 from api.db import db
 from api.config import config
+from api.app import create_app
```
Comment (yashLadha) [inline backend/blueprint/api/run.py]: Won't it be 
```py
from api import create_app
```

**Comment by yashLadha** [comment]
@vaibhavsingh97 there is no such package as `app` in the code and you are using the `__init__` file of the api package so please test the functions before commiting.

File: backend/blueprint/api/app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import os
```
Comment (yashLadha) [inline backend/blueprint/api/app/__init__.py]: Why to move in a different folder?

**Comment by yashLadha** [comment]
Have a look at this link it and particularly at __Step 5__ it will give you a better idea about what i am proposing.

File: backend/blueprint/api/app/__init__.py (hunk 2 above)
Comment (vaibhavsingh97) [reply backend/blueprint/api/app/__init__.py]: Ok, you are right, I am correcting this. Thanks for pointing it out

**Comment by vaibhavsingh97** [comment]
@yashLadha I had added the suggested changes. Please review it again

**Comment by yashLadha** [comment]
@vaibhavsingh97 have you seen the [link ](https://www.digitalocean.com/community/tutorials/how-to-structure-large-flask-applications) Please try to follow the same as the standard.

**Comment by vaibhavsingh97** [comment]
@yashLadha @ParthS007 I followed that link, I found everything correct. can you please point out where I am wrong, so that I can change it

**Comment by yashLadha** [comment]
As you can see `run.py` should be just run the server and should only import app to run, but you are instantiating config in the `run.py` and that should not happen.

**Comment by yashLadha** [comment]
Sure @djmgit

**Comment by vaibhavsingh97** [comment]
@yashLadha 
You asked me to look into https://www.digitalocean.com/community/tutorials/how-to-structure-large-flask-applications
I had a look at `Step 5`
image

I am also doing the same
image

So here what I am doing is instead of loading whole config file, I am loading config according to the type of environment i.e `Development | Production | Testing | Staging`

Can you please tell me what's wrong in here?

Cc: @djmgit your views

**Comment by yashLadha** [comment]
Please look at this PR to better understand the concept @vaibhavsingh97 https://github.com/fossasia/badgeyay/pull/765

**Comment by vaibhavsingh97** [comment]
@yashLadha Done with the changes, please review it again

File: backend/blueprint/api/run.py [hunk 3]
```diff
@@ -37,6 +26,8 @@
 app.register_blueprint(errorHandlers.router)
 app.register_blueprint(resetUser.router, url_prefix='/reset')
 
+
+
```
Comment (ParthS007) [inline backend/blueprint/api/run.py]: Please leave only 2 lines :+1:

File: backend/blueprint/api/controllers/homePage.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 router = Blueprint('homePage', __name__)
 
 
-@router.route('/', methods=['POST'])
+@router.route('/', methods=['GET', 'POST'])
```
Comment (ParthS007) [inline backend/blueprint/api/controllers/homePage.py]: Can you please explain why we need `GET` here?

File: backend/app/requirements.txt [hunk 5]
```diff
@@ -1,21 +1,33 @@
...
+Pillow==5.1.0
+psycopg2-binary==2.7.4
+pycparser==2.18
+pyflakes==1.6.0
 PyJWT==1.4.2
-Flask-SQLAlchemy==2.1
-Flask-Mail==0.9.1
-
+PyPDF2==1.26.0
+python-dateutil==2.7.2
```
Comment (ParthS007) [inline backend/app/requirements.txt]: I think It is not used in our project.

File: backend/app/requirements.txt [hunk 6]
```diff
@@ -1,21 +1,33 @@
...
+python-dotenv==0.8.2
+python-editor==1.0.3
+pytz==2018.4
+PyYAML==3.12
```
Comment (ParthS007) [inline backend/app/requirements.txt]: Same

Comment (ParthS007) [reply backend/app/requirements.txt]: Remove those which are not needed :+1:

Comment (vaibhavsingh97) [reply backend/app/requirements.txt]: I haven't added anything, it's sub-dependencies. I just did `pip freeze > requirements.txt` to add the installed libraries

File: backend/app/requirements.txt (hunk 5 above)
Comment (vaibhavsingh97) [reply backend/app/requirements.txt]: I haven't added anything, it's sub-dependencies. I just did `pip freeze > requirements.txt` to add the installed libraries

**Comment by yashLadha** [comment]
What is the status of the PR? Please make the requested changes by @ParthS007

**Comment by yashLadha** [comment]
Holding the PR for such a long time is not a development practice and you should submit the Patch for the changes immediately as requested, else the PR is good to be closed.

**Comment by ParthS007** [comment]
What's the status of this PR @vaibhavsingh97?
This PR can be closed as most of the changes are done already and it is not updated for so long.
Please follow best practices.
cheers !!

**Comment by vaibhavsingh97** [comment]
@ParthS007 Yes, I agreed most of the changes are done, But I don't know why most of the PR gets merged before this PR, Hence, making this PR irrelevant. But this PR still increases the modularity, so I will fix conflicts

**Comment by yashLadha** [comment]
Please resolve the travis error and conflicts

File: backend/blueprint/api/controllers/homePage.py (hunk 4 above)
Comment (vaibhavsingh97) [reply backend/blueprint/api/controllers/homePage.py]: I added `GET` to test, removing 👍

**Comment by vaibhavsingh97** [comment]
@djmgit @yashLadha @gabru-md @ParthS007 PLease review

File: backend/blueprint/api/__init__.py [hunk 7]
```diff
@@ -19,6 +25,9 @@ def create_app():
     app.config['SECRET_KEY'] = config.POSTGRES['secret']
     app.config['DEBUG'] = config.DEBUG
     app.config.from_object('api.config.mailConfig.MailConfig')
+    app.config.from_object(os.getenv('APP_CONFIG', default='api.config.config.ProductionConfig'))
     CORS(app, resources={r"*": {"origins": "*"}})
+    db.init_app(app)
+    Migrate(app, db)
```
Comment (yashLadha) [inline backend/blueprint/api/__init__.py]: What is the use case of shifting it here?? :confused:

Comment (vaibhavsingh97) [reply backend/blueprint/api/__init__.py]: Because it should be defined here only

Comment (vaibhavsingh97) [reply backend/blueprint/api/__init__.py]: Moreover, `open-event-server` follows the same approach https://github.com/fossasia/open-event-server/blob/a80b40449ccbfb183d6fb2680fdb36cb53c4ba05/app/__init__.py#L72

Comment (yashLadha) [reply backend/blueprint/api/__init__.py]: Them you probably also have seen the blueprint registration, also add the same. Cheers

Comment (vaibhavsingh97) [reply backend/blueprint/api/__init__.py]: @yashLadha , Sure I will update, thanks

Comment (vaibhavsingh97) [reply backend/blueprint/api/__init__.py]: Done @yashLadha 👍

File: backend/app/requirements.txt (hunk 6 above)
Comment (yashLadha) [reply backend/app/requirements.txt]: It installs the dependencies that are downloaded by other dependencies as well. This makes the file very big and less readable as the project develops. Try to just include the libraries that are absolutely necessary and not `freeze`

**Comment by vaibhavsingh97** [comment]
@yashLadha It is good to have all packages in the `requirements.txt` because one is sure about what dependencies versions work with your environment.
Think about some package getting updated, and breaking backwards compatibilities: one would have problems.

**Comment by yashLadha** [comment]
Them those things you have to do in the development as well and at that time you can update the same. I don't think it takes so much importance and why are you persistent about managing dependencies when we have other issues to tackle.

**Comment by yashLadha** [comment]
And I haven't seen any python project to not maintain requirements.txt and not use virtualenv. It is just matter of preference. Hope you understand that right now priorities are to make the website in stable state and can you help us doing the same. That would be of great help, cheers

**Comment by vaibhavsingh97** [comment]
@djmgit @gabru-md @ParthS007 @yashLadha can you please review this PR. i will fix conflicts and add changes after your suggestions. thanks

**Comment by gabru-md** [comment]
Closing, Since it is not required

**Pull request closed without merging**